content/en/docs/release-oversight/testing-in-production.md

```diff
@@ -0,0 +1,45 @@
+
+### Our philosophy on failure
+It's important to understand we believe breaking ourselves first is a form of success.  Failing a sprint is a gift in this sense.  It gives us time to adjust before our next big release.  What this means in practice is that, when sprints fail, we commit to clearing our sprint of feature work and focusing on stability across whatever teams necessary.
+
+### Sprint success or failure
+There are couple ways we judge the success or a failure of a sprint:
+
+* Did we have a build on time?
+* Were any changes required by the administrator not covered by an [admin ack](https://docs.openshift.com/container-platform/4.9/updating/updating-cluster-prepare.html#update-preparing-migrate_updating-cluster-prepare)?
+* Did we find a significant problem that forced us to remove the production cluster from rotation?
+
+### Selecting the build
+The selection process is quite simple.  On the last Wednesday of each sprint the [Technical Release Team](/docs/release-oversight/the-technical-release-team/) (TRT) will select an accepted payload from the last two days.  From there Quality Engineering will run additional tests.  Barring any new surprises Test Platform will be notified that a release is ready.
+
+It's worth noting that we'll likely create a dedicated channel for managing these releases through the [Update Service](https://docs.openshift.com/container-platform/4.7/updating/understanding-the-update-service.html) in the upcoming months.
+
+### The environment
+We've selected one of our build farm clusters to run the output of each sprint.  This is a critical piece of infrastructure that all development teams are motivated to keep working.  In this document we'll refer to this cluster as the *sacrificial cluster*.
+
+### Investigating problems
+The TRT has admin access to the sacrificial cluster for debugging purposes.  As needed this access can be temporarily granted by Test Platform to any other development team that needs it.  Temporary configuration changes are allowed for experiments but anything needed to persist will require failing the sprint.
+
+### Hotfixes
+Hotfixes are possible but not desired.  Changes for the sole purpose of gathering more data are preferred over changes in functionality.  Before we resort to patching a problem we'll default to [reverting the change](/docs/release-oversight/improving-ci-signal/).  Changes like this are best done as part of the subsequent sprint release.
+
+### When things go wrong the sacrificial cluster is removed from rotation
+The Test Platform team maintains the infrastructure that powers our CI builds.  When there are major outages they already have an incident management process in place.  What's different now is that the TRT can detect subtle degradations as a result of sprint upgrades.  In most of these cases the TRT contacts Test Platform via their helpdesk bot to request the sacrificial cluster be removed from rotation.  Behind the scenes a Jira is created for tracking.
```

Comment:
My impression was that test-platform would be watching for signs of degradation, or hearing about degradation from the CI-consuming devs.  And then they'd hand the cluster off to TRT or updates or otherwise for inspection/debugging, and possibly fail over the CI load to other clusters.  That makes sense to me, because CI is TP's baby, and they are well positioned to confirm whether it is healthy or sad.  But as long as the sacrificial cluster is hosting a bunch of 4.dev release blocking jobs, TRT will be one of the CI-consumers that are likely to notice too, without needing deep insight into the TP CI tooling.  So either way...

Reply:
This MR is based on a discussion with Test Platform yesterday.  I probably need to make it more clear, but the general idea is that TP (and likely many other teams) will notice major problems and remove the sacrificial cluster from rotation.  The TRT is able to focus on nearly unperceivable degradations.  Since they are the only team focusing on these they will request request the failover whenever necessary for further problem isolation.  Removing the sacrificial cluster from rotation does not imply you are on the hook for bringing it back in.  The TRT is most incentivized to get the cluster back in therefore they'll drive incident management (not necessarily fixing everything).

Reply:
Could we clarify more here what happens if removed from rotation? Is it out of rotation for 3 weeks until the next attempt? Let say we find something to revert, or there's a hotfix that seems to work, it seems like it would be good to get it back into rotation, but this implies doing builds of "bad sprintly build + hotfix" somehow. 

If we do leave it out for 3 weeks, there's risk there was a secondary problem we couldn't see because of the primary problem, which now starts surfacing when the next build lands.

Reply:
I can certainly try! :smile: 

My feeling is that it's going to be a case by case situation.  I think sometimes when it's removed from rotation we aren't going to find any evidence that the sacrificial cluster was actually the problem.  Other times we may learn the problem isn't as bad as we thought or we have some way to mitigate it.

At this point I rather wait and see if we can really avoid hotfixes.  I liked your suggestion of the gitops model of 

1. move the system into rescue mode
2. hack it up to make it work
3. undo everything at the end as we upgrade to the next sprint to make sure we get back to a known state 

I can't say I know exactly how that will work on technical terms though.  If someone here thinks they could explain it you are welcome to add an update. :smile:

Reply:
Check the updates I just pushed and let me know what you think.

Reply:
When the cluster is removed from the rotation , it does not have to wait for end of next sprint to get fixed. This issue should become a blocker issue for the team which need to fix the issue. We should add a sentence that `More time the cluster is out of the rotation  we are at risk of not detecting some other issue. So we need to get the cluster up and running as soon as possible.`

Reply:
I added "The goal in this case is for the sacrificial cluster to spend as much time in rotation as possible and hopefully catch any problems that could put the next sprint at risk." to the "Reasonable effort will be made to bring the sacrificial cluster back into rotation" section.